File: typescript/src/core/transport.ts
/**
 * The low-level HTTP transport: header/auth assembly, retries with backoff,
 * JSON and multipart requests, binary responses, and SSE streaming.
 *
 * Depends only on the global `fetch` (Node >= 18, browsers, edge runtimes); a
 * custom `fetch` can be injected for testing or non-standard runtimes.
 */

import {
  AirforceConnectionError,
  AirforceError,
  AirforceTimeoutError,
  MissingCredentialError,
} from "./errors";
import { Stream } from "./streaming";

export const VERSION = "0.0.1";
export const DEFAULT_BASE_URL = "https://api.airforce";

/** Which credential a given endpoint expects. */
export type AuthMode = "api_key" | "session" | "none";

export type QueryValue = string | number | boolean | undefined | null;

export interface RequestOptions {
  method: "GET" | "POST" | "PUT" | "PATCH" | "DELETE";
  path: string;
  /** Which credential to attach. Defaults to `api_key`. */
  auth?: AuthMode;
  query?: Record<string, QueryValue>;
  /** JSON body. Mutually exclusive with `form` / `urlencoded`. */
  body?: unknown;
  /** Multipart form body. Mutually exclusive with `body` / `urlencoded`. */
  form?: FormData;
  /** `application/x-www-form-urlencoded` body (OAuth token/revoke). */
  urlencoded?: Record<string, string | undefined>;
  headers?: Record<string, string>;
  /** Per-request timeout override (ms). */
  timeout?: number;
  /** Per-request retry override. */
  maxRetries?: number;
  /** Caller-supplied abort signal. */
  signal?: AbortSignal;
}

export interface TransportConfig {
  apiKey?: string;
  sessionToken?: string;
  baseURL: string;
  timeout: number;
  maxRetries: number;
  defaultHeaders: Record<string, string>;
  fetch: typeof fetch;
}

// 409 is intentionally excluded: it is a terminal business conflict (e.g. "already
// subscribed" on checkout), not a transient error worth retrying.
const RETRYABLE_STATUS = new Set([408, 429, 500, 502, 503, 504]);

interface FetchResponseOutcome {
  kind: "response";
  response: Response;
}
interface FetchErrorOutcome {
  kind: "error";
  error: unknown;
  timedOut: boolean;
}
type FetchOutcome = FetchResponseOutcome | FetchErrorOutcome;

function describeError(error: unknown): string {
  if (error instanceof Error) return error.message;
  if (typeof error === "string") return error;
  try {
    return JSON.stringify(error) ?? "unknown error";
  } catch {
    return "unknown error";
  }
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted) return reject(new AirforceError("Aborted"));
    const t = setTimeout(resolve, ms);
    signal?.addEventListener(
      "abort",
      () => {
        clearTimeout(t);
        reject(new AirforceError("Aborted"));
      },
      { once: true },
    );
  });
}

export class Transport {
  constructor(private readonly config: TransportConfig) {}

  /** Update the session token (e.g. after `auth.login()`). */
  setSessionToken(token: string | undefined): void {
    this.config.sessionToken = token;
  }

  /** Update the API key. */
  setApiKey(key: string | undefined): void {
    this.config.apiKey = key;
  }

  /** Perform a request and decode the JSON response into `T`. */
  async request<T>(options: RequestOptions): Promise<T> {
    return (await this.requestDetailed<T>(options)).data;
  }

  /** Like {@link request} but also returns the response headers. */
  async requestDetailed<T>(
    options: RequestOptions,
  ): Promise<{ data: T; headers: Headers }> {
    const res = await this.send(options, false);
    const headers = res.headers;
    if (res.status === 204) return { data: undefined as T, headers };
    const text = await res.text();
    if (!text) return { data: undefined as T, headers };
    try {
      return { data: JSON.parse(text) as T, headers };
    } catch {
      // Some endpoints return a bare string body.
      return { data: text as unknown as T, headers };
    }
  }

  /** Perform a request and return the raw binary body (audio, etc.). */
  async requestBinary(options: RequestOptions): Promise<ArrayBuffer> {
    const res = await this.send(options, false);
    return res.arrayBuffer();
  }

  /** Perform a streaming request and return a typed SSE {@link Stream}. */
  async stream<T>(options: RequestOptions): Promise<Stream<T>> {
    const controller = new AbortController();
    if (options.signal) {
      options.signal.addEventListener("abort", () => controller.abort(), {
        once: true,
      });
    }
    const res = await this.send(
      { ...options, signal: controller.signal },
      true,
    );
    if (!res.body) {
      throw new AirforceError("Streaming response had no body");
    }
    return Stream.fromSSE<T>(res.body, controller);
  }

  // --- internals -----------------------------------------------------------

  private resolveToken(auth: AuthMode): string | undefined {
    if (auth === "none") return undefined;
    // Session endpoints require a session JWT — never silently substitute an API
    // key (that would send the wrong credential and mask MissingCredentialError).
    if (auth === "session") return this.config.sessionToken;
    return this.config.apiKey ?? this.config.sessionToken;
  }

  private buildURL(path: string, query?: Record<string, QueryValue>): string {
    const url = new URL(
      path.startsWith("/") ? path.slice(1) : path,
      this.config.baseURL.endsWith("/")
        ? this.config.baseURL
        : this.config.baseURL + "/",
    );
    if (query) {
      for (const [k, v] of Object.entries(query)) {
        if (v !== undefined && v !== null) url.searchParams.set(k, String(v));
      }
    }
    return url.toString();
  }

  private buildHeaders(
    options: RequestOptions,
    streaming: boolean,
  ): Headers {
    const headers = new Headers(this.config.defaultHeaders);
    headers.set("user-agent", `airforce-sdk-js/${VERSION}`);
    headers.set("x-airforce-sdk", `js/${VERSION}`);
    headers.set("accept", streaming ? "text/event-stream" : "application/json");

    const auth: AuthMode = options.auth ?? "api_key";
    const token = this.resolveToken(auth);
    if (auth !== "none") {
      if (!token) {
        throw new MissingCredentialError(
          auth === "session"
            ? "This endpoint requires a session token (set `sessionToken`, e.g. from auth.login())."
            : "This endpoint requires an API key (set `apiKey`).",
          { code: "missing_credential" },
        );
      }
      headers.set("authorization", `Bearer ${token}`);
    }

    if (options.urlencoded) {
      headers.set("content-type", "application/x-www-form-urlencoded");
    } else if (options.body !== undefined && !options.form) {
      headers.set("content-type", "application/json");
    }
    if (options.headers) {
      for (const [k, v] of Object.entries(options.headers)) headers.set(k, v);
    }
    return headers;
  }

  private buildRequestInit(
    options: RequestOptions,
    streaming: boolean,
  ): RequestInit {
    const init: RequestInit = {
      method: options.method,
      headers: this.buildHeaders(options, streaming),
    };
    if (options.form) {
      init.body = options.form; // fetch sets the multipart boundary
    } else if (options.urlencoded) {
      const params = new URLSearchParams();
      for (const [k, v] of Object.entries(options.urlencoded)) {
        if (v !== undefined) params.set(k, v);
      }
      init.body = params;
    } else if (options.body !== undefined) {
      init.body = JSON.stringify(options.body);
    }
    return init;
  }

  /** Run a single attempt under a timeout, classifying the outcome. */
  private async attemptFetch(
    url: string,
    init: RequestInit,
    timeout: number,
    signal: AbortSignal | undefined,
  ): Promise<FetchOutcome> {
    const timeoutController = new AbortController();
    const timer = setTimeout(() => timeoutController.abort(), timeout);
    const signals = signal
      ? [timeoutController.signal, signal]
      : [timeoutController.signal];
    try {
      const response = await this.config.fetch(url, {
        ...init,
        signal: anySignal(signals),
      });
      return { kind: "response", response };
    } catch (error) {
      return { kind: "error", error, timedOut: timeoutController.signal.aborted };
    } finally {
      clearTimeout(timer);
    }
  }

  private async send(
    options: RequestOptions,
    streaming: boolean,
  ): Promise<Response> {
    const url = this.buildURL(options.path, options.query);
    const init = this.buildRequestInit(options, streaming);
    const maxRetries = options.maxRetries ?? this.config.maxRetries;
    const timeout = options.timeout ?? this.config.timeout;

    for (let attempt = 0; ; attempt++) {
      const canRetry = attempt < maxRetries;
      const outcome = await this.attemptFetch(url, init, timeout, options.signal);
      const res =
        outcome.kind === "response"
          ? await this.handleResponse(outcome.response, attempt, canRetry, options.signal)
          : await this.handleError(outcome, options, url, timeout, attempt, canRetry);
      if (res) return res;
      // res === null means a retry was scheduled — loop again.
    }
  }

  /** Resolve a successful attempt: return the Response, retry, or throw. */
  private async handleResponse(
    res: Response,
    attempt: number,
    canRetry: boolean,
    signal: AbortSignal | undefined,
  ): Promise<Response | null> {
    if (res.ok) return res;
    if (RETRYABLE_STATUS.has(res.status) && canRetry) {
      await this.backoff(attempt, parseRetryAfter(res.headers), signal);
      return null;
    }
    throw AirforceError.fromResponse(res.status, await safeJson(res), res.headers);
  }

  /** Resolve a failed attempt: retry (return null) or throw a typed error. */
  private async handleError(
    outcome: FetchErrorOutcome,
    options: RequestOptions,
    url: string,
    timeout: number,
    attempt: number,
    canRetry: boolean,
  ): Promise<null> {
    if (options.signal?.aborted) {
      throw new AirforceError("Aborted", { cause: outcome.error });
    }
    // A network/timeout error leaves the outcome of a POST unknown — retrying could
    // double-charge a billable request. Only retry transport errors for idempotent
    // methods. (Retryable *status codes* are still retried for all methods, since the
    // server responded and a 4xx/5xx POST is not charged.)
    if (canRetry && options.method !== "POST") {
      await this.backoff(attempt, undefined, options.signal);
      return null;
    }
    throw outcome.timedOut
      ? new AirforceTimeoutError(
          `Request to ${options.path} timed out after ${timeout}ms`,
          outcome.error,
        )
      : new AirforceConnectionError(
          `Failed to reach ${url}: ${describeError(outcome.error)}`,
          outcome.error,
        );
  }

  private async backoff(
    attempt: number,
    retryAfterSecs: number | undefined,
    signal?: AbortSignal,
  ): Promise<void> {
    const base =
      retryAfterSecs !== undefined
        ? retryAfterSecs * 1000
        : Math.min(1000 * 2 ** attempt, 8000);
    const jitter = base * 0.25 * Math.random();
    await sleep(base + jitter, signal);
  }
}

function parseRetryAfter(headers: Headers): number | undefined {
  const v = headers.get("retry-after");
  if (!v) return undefined;
  const secs = Number(v);
  return Number.isFinite(secs) ? secs : undefined;
}

async function safeJson(res: Response): Promise<unknown> {
  const text = await res.text().catch(() => "");
  if (!text) return undefined;
  try {
    return JSON.parse(text);
  } catch {
    return text;
  }
}

/** Combine multiple abort signals into one (AbortSignal.any polyfill). */
function anySignal(signals: AbortSignal[]): AbortSignal {
  if (signals.length === 1) return signals[0]!;
  const anyFn = (AbortSignal as unknown as { any?: (s: AbortSignal[]) => AbortSignal }).any;
  if (typeof anyFn === "function") return anyFn(signals);
  const controller = new AbortController();
  for (const s of signals) {
    if (s.aborted) {
      controller.abort();
      break;
    }
    s.addEventListener("abort", () => controller.abort(), { once: true });
  }
  return controller.signal;
}

File: java/src/main/java/com/airforce/Transport.java
package com.airforce;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpTimeoutException;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ThreadLocalRandom;

class Transport {

  static final String VERSION = "0.0.1";
  // 409 is excluded: a terminal business conflict (e.g. "already subscribed"), not transient.
  private static final Set<Integer> RETRYABLE = Set.of(408, 429, 500, 502, 503, 504);

  private final HttpClient http;
  private final ObjectMapper mapper = new ObjectMapper();
  private final String baseUrl;
  private final int maxRetries;
  private final Duration timeout;
  private final Map<String, String> defaultHeaders;

  private String apiKey;
  private String sessionToken;

  Transport(HttpClient http, String apiKey, String sessionToken, String baseUrl,
      Duration timeout, int maxRetries, Map<String, String> defaultHeaders) {
    this.http = http;
    this.apiKey = apiKey;
    this.sessionToken = sessionToken;
    this.baseUrl = baseUrl;
    this.timeout = timeout;
    this.maxRetries = maxRetries;
    this.defaultHeaders = defaultHeaders;
  }

  ObjectMapper mapper() {
    return mapper;
  }

  void setSessionToken(String token) {
    this.sessionToken = token;
  }

  String sessionToken() {
    return sessionToken;
  }

  String baseUrl() {
    return baseUrl;
  }

  // --- request spec --------------------------------------------------------

  static class RequestSpec {
    String auth = "api_key";
    Map<String, String> query;
    Object jsonBody;
    byte[] rawBody;
    String contentType;
    Map<String, String> headers;
  }

  // --- public entry points -------------------------------------------------

  JsonNode requestJson(String method, String path, RequestSpec spec) {
    HttpResponse<InputStream> resp = send(method, path, spec, false);
    try (InputStream in = resp.body()) {
      byte[] bytes = in.readAllBytes();
      if (resp.statusCode() == 204 || bytes.length == 0) {
        return mapper.nullNode();
      }
      return mapper.readTree(bytes);
    } catch (IOException e) {
      throw new AirforceException.ApiConnection("airforce: failed reading response", e);
    }
  }

  /** Holder for a JSON response plus the airforce_session cookie (auth flow). */
  static class JsonAndCookie {
    final JsonNode json;
    final String sessionCookie;

    JsonAndCookie(JsonNode json, String sessionCookie) {
      this.json = json;
      this.sessionCookie = sessionCookie;
    }
  }

  JsonAndCookie requestJsonCookie(String method, String path, RequestSpec spec) {
    HttpResponse<InputStream> resp = send(method, path, spec, false);
    JsonNode json;
    try (InputStream in = resp.body()) {
      byte[] bytes = in.readAllBytes();
      json = bytes.length == 0 ? mapper.nullNode() : mapper.readTree(bytes);
    } catch (IOException e) {
      throw new AirforceException.ApiConnection("airforce: failed reading response", e);
    }
    String cookie = null;
    for (String header : resp.headers().allValues("set-cookie")) {
      int idx = header.indexOf("airforce_session=");
      if (idx >= 0) {
        String rest = header.substring(idx + "airforce_session=".length());
        int semi = rest.indexOf(';');
        cookie = semi >= 0 ? rest.substring(0, semi) : rest;
        break;
      }
    }
    return new JsonAndCookie(json, cookie);
  }

  byte[] requestBytes(String method, String path, RequestSpec spec) {
    HttpResponse<InputStream> resp = send(method, path, spec, false);
    try (InputStream in = resp.body()) {
      return in.readAllBytes();
    } catch (IOException e) {
      throw new AirforceException.ApiConnection("airforce: failed reading response", e);
    }
  }

  Stream requestStream(String method, String path, RequestSpec spec) {
    HttpResponse<InputStream> resp = send(method, path, spec, true);
    return new Stream(resp.body(), mapper);
  }

  // convenience wrappers used by resources

  JsonNode method(String method, String path, String auth, Object body) {
    RequestSpec spec = new RequestSpec();
    spec.auth = auth;
    spec.jsonBody = body;
    return requestJson(method, path, spec);
  }

  JsonNode post(String path, String auth, Object body) {
    return method("POST", path, auth, body);
  }

  JsonNode get(String path, String auth, Map<String, String> query) {
    RequestSpec spec = new RequestSpec();
    spec.auth = auth;
    spec.query = query;
    return requestJson("GET", path, spec);
  }

  JsonNode delete(String path, String auth) {
    RequestSpec spec = new RequestSpec();
    spec.auth = auth;
    return requestJson("DELETE", path, spec);
  }

  byte[] postBytes(String path, String auth, Object body) {
    RequestSpec spec = new RequestSpec();
    spec.auth = auth;
    spec.jsonBody = body;
    return requestBytes("POST", path, spec);
  }

  byte[] getBytes(String path, String auth) {
    RequestSpec spec = new RequestSpec();
    spec.auth = auth;
    return requestBytes("GET", path, spec);
  }

  Stream postStream(String path, String auth, Object body) {
    RequestSpec spec = new RequestSpec();
    spec.auth = auth;
    spec.jsonBody = body;
    return requestStream("POST", path, spec);
  }

  // --- internals -----------------------------------------------------------

  private String resolveToken(String auth) {
    if ("none".equals(auth)) {
      return null;
    }
    // Session endpoints require a session JWT — never substitute an API key.
    String token = "session".equals(auth)
        ? sessionToken
        : firstNonEmpty(apiKey, sessionToken);
    if (token == null) {
      throw new AirforceException.MissingCredential(
          "session".equals(auth)
              ? "This endpoint requires a session token (set sessionToken / auth().login())."
              : "This endpoint requires an API key (set apiKey).");
    }
    return token;
  }

  private byte[] bodyBytes(RequestSpec spec) {
    if (spec.rawBody != null) {
      return spec.rawBody;
    }
    if (spec.jsonBody != null) {
      try {
        return mapper.writeValueAsBytes(spec.jsonBody);
      } catch (IOException e) {
        throw new AirforceException("airforce: failed to encode request body", 0, null, null, null, 0, null, e);
      }
    }
    return null;
  }

  private HttpRequest buildRequest(String method, String url, RequestSpec spec, String token,
      String contentType, byte[] body, boolean stream) {
    HttpRequest.BodyPublisher publisher = body != null
        ? HttpRequest.BodyPublishers.ofByteArray(body)
        : HttpRequest.BodyPublishers.noBody();
    HttpRequest.Builder b = HttpRequest.newBuilder(URI.create(url))
        .timeout(timeout)
        .method(method, publisher)
        .header("user-agent", "airforce-sdk-java/" + VERSION)
        .header("x-airforce-sdk", "java/" + VERSION)
        .header("accept", stream ? "text/event-stream" : "application/json");
    if (contentType != null) {
      b.header("content-type", contentType);
    }
    if (token != null) {
      b.header("authorization", "Bearer " + token);
    }
    defaultHeaders.forEach(b::header);
    if (spec.headers != null) {
      spec.headers.forEach(b::header);
    }
    return b.build();
  }

  private HttpResponse<InputStream> send(String method, String path, RequestSpec spec, boolean stream) {
    String url = buildUrl(path, spec.query);
    String token = resolveToken(spec.auth);
    byte[] body = bodyBytes(spec);
    String contentType = spec.contentType != null
        ? spec.contentType
        : (spec.jsonBody != null ? "application/json" : null);

    for (int attempt = 0; ; attempt++) {
      HttpRequest request = buildRequest(method, url, spec, token, contentType, body, stream);
      HttpResponse<InputStream> resp = attempt(request, path, attempt);
      if (resp == null) {
        continue; // a retry was scheduled
      }
      int code = resp.statusCode();
      if (code < 400) {
        return resp;
      }
      if (RETRYABLE.contains(code) && attempt < maxRetries) {
        double retryAfter = retryAfterSeconds(resp);
        drain(resp);
        sleep(backoffMillis(attempt + 1, retryAfter));
        continue;
      }
      throw AirforceException.fromResponse(code, readBody(resp),
          resp.headers().firstValue("x-request-id").orElse(null), retryAfterSeconds(resp));
    }
  }

  /** One HTTP attempt; returns null (after scheduling a retry) or throws on terminal failure. */
  private HttpResponse<InputStream> attempt(HttpRequest request, String path, int attempt) {
    // A transport error leaves a POST's outcome unknown — retrying could double-charge a
    // billable request. Only retry idempotent methods.
    boolean idempotent = !"POST".equalsIgnoreCase(request.method());
    try {
      return http.send(request, HttpResponse.BodyHandlers.ofInputStream());
    } catch (HttpTimeoutException e) {
      if (attempt < maxRetries && idempotent) {
        sleep(backoffMillis(attempt + 1, 0));
        return null;
      }
      throw new AirforceException.ApiTimeout("airforce: request to " + path + " timed out", e);
    } catch (IOException e) {
      if (attempt < maxRetries && idempotent) {
        sleep(backoffMillis(attempt + 1, 0));
        return null;
      }
      throw new AirforceException.ApiConnection("airforce: request to " + path + " failed: " + e.getMessage(), e);
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new AirforceException.ApiConnection("airforce: request interrupted", e);
    }
  }

  private String buildUrl(String path, Map<String, String> query) {
    StringBuilder url = new StringBuilder(baseUrl);
    url.append(path.startsWith("/") ? path : "/" + path);
    if (query != null && !query.isEmpty()) {
      StringBuilder qs = new StringBuilder();
      for (Map.Entry<String, String> e : query.entrySet()) {
        if (e.getValue() == null) {
          continue;
        }
        qs.append(qs.length() == 0 ? '?' : '&')
            .append(URLEncoder.encode(e.getKey(), StandardCharsets.UTF_8))
            .append('=')
            .append(URLEncoder.encode(e.getValue(), StandardCharsets.UTF_8));
      }
      url.append(qs);
    }
    return url.toString();
  }

  private static String firstNonEmpty(String a, String b) {
    if (a != null && !a.isEmpty()) {
      return a;
    }
    return (b != null && !b.isEmpty()) ? b : null;
  }

  private static double retryAfterSeconds(HttpResponse<?> resp) {
    return resp.headers().firstValue("retry-after").map(v -> {
      try {
        return Double.parseDouble(v);
      } catch (NumberFormatException e) {
        return 0.0;
      }
    }).orElse(0.0);
  }

  private static long backoffMillis(int attempt, double retryAfterSeconds) {
    double base = retryAfterSeconds > 0 ? retryAfterSeconds : Math.min(Math.pow(2, attempt - 1), 8);
    double jitter = base * 0.25 * ThreadLocalRandom.current().nextDouble();
    return (long) ((base + jitter) * 1000);
  }

  private static void sleep(long millis) {
    try {
      Thread.sleep(millis);
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new AirforceException.ApiConnection("airforce: retry sleep interrupted", e);
    }
  }

  private static void drain(HttpResponse<InputStream> resp) {
    try (InputStream in = resp.body()) {
      in.readAllBytes();
    } catch (IOException ignored) {
      // best effort
    }
  }

  private static String readBody(HttpResponse<InputStream> resp) {
    try (InputStream in = resp.body()) {
      return new String(in.readAllBytes(), StandardCharsets.UTF_8);
    } catch (IOException e) {
      return "";
    }
  }
}
